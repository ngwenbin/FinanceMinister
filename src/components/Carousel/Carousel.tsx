import {
  Dimensions,
  FlatList,
  View,
  NativeSyntheticEvent,
  NativeScrollEvent,
  FlatListProps,
  ViewStyle,
  ColorValue,
} from "react-native";
import { memo, useCallback, useRef, useState } from "react";
import { ArrayElement } from "@utils/tsHelpers";
import { isEqual } from "lodash";
import { scaledStylesheet } from "@utils";
import { Colors } from "@constants/Colors";

const { width: windowWidth } = Dimensions.get("window");

interface PaginationProps<T> {
  data: Array<T>;
  index: number;
  style?: ViewStyle;
  activeColor?: ColorValue;
  inActiveColor?: ColorValue;
}

interface CarouselProps<T> {
  data: Array<T & { id: string }>;
  containerContent: (item: T) => JSX.Element;
  initialNumToRender?: number;
  maxToRenderPerBatch?: number;
  windowSize?: number;
  showPagination?: boolean;
  containerStyle?: ViewStyle;
  paginationStyle?: ViewStyle;
  activePaginationColor?: ColorValue;
  inActivePaginationColor?: ColorValue;
}

const Pagination = <T,>({
  data,
  index,
  style,
  activeColor = Colors.gray[700],
  inActiveColor = Colors.gray[300],
}: PaginationProps<T>) => {
  return (
    <View style={styles.pagination} pointerEvents="none">
      {data.map((_, i) => {
        return (
          <View
            key={i}
            style={[
              styles.paginationDot,
              {
                backgroundColor: index === i ? activeColor : inActiveColor,
              },
              style,
            ]}
          />
        );
      })}
    </View>
  );
};

const memoEqualityCheck = (prev: any, next: any) => isEqual(prev, next);

const Slider = memo(
  ({ item, renderer }: { item: any; renderer: (data: any) => JSX.Element }) => {
    return renderer(item);
  },
  memoEqualityCheck
);

const Carousel = <T,>({
  data,
  containerContent,
  initialNumToRender = 0,
  maxToRenderPerBatch = 1,
  windowSize = 2,
  showPagination = true,
  containerStyle,
  paginationStyle,
  activePaginationColor,
  inActivePaginationColor,
}: CarouselProps<T>) => {
  const [index, setIndex] = useState(0);
  const indexRef = useRef(index);
  indexRef.current = index;
  const onScroll = useCallback(
    (event: NativeSyntheticEvent<NativeScrollEvent>) => {
      const slideSize = event.nativeEvent.layoutMeasurement.width;
      const index = event.nativeEvent.contentOffset.x / slideSize;
      const roundIndex = Math.round(index);

      const distance = Math.abs(roundIndex - index);
      const isNoMansLand = distance > 0.4;

      if (roundIndex !== indexRef.current && !isNoMansLand) {
        setIndex(roundIndex);
      }
    },
    []
  );

  const flatListOptimizationProps: Partial<
    FlatListProps<ArrayElement<typeof data>>
  > = {
    initialNumToRender,
    maxToRenderPerBatch,
    removeClippedSubviews: true,
    scrollEventThrottle: 16,
    windowSize,
    keyExtractor: useCallback(
      (item: ArrayElement<typeof data>) => String(item.id),
      []
    ),
    getItemLayout: useCallback(
      (_: any, index: number) => ({
        index,
        length: windowWidth,
        offset: index * windowWidth,
      }),
      []
    ),
  };

  return (
    <View style={containerStyle}>
      <FlatList
        data={data}
        style={{ flex: 1 }}
        renderItem={({ item }) => (
          <Slider item={item} renderer={containerContent} />
        )}
        pagingEnabled
        horizontal
        showsHorizontalScrollIndicator={false}
        bounces={false}
        onScroll={onScroll}
        {...flatListOptimizationProps}
      />
      {showPagination ? (
        <Pagination
          data={data}
          index={index}
          style={paginationStyle}
          activeColor={activePaginationColor}
          inActiveColor={inActivePaginationColor}
        />
      ) : null}
    </View>
  );
};

export default Carousel;

const styles = scaledStylesheet({
  pagination: {
    paddingVertical: 2,
    width: "100%",
    justifyContent: "center",
    flexDirection: "row",
  },
  paginationDot: {
    width: 6,
    height: 6,
    borderRadius: 10,
    marginHorizontal: 2,
  },
});
