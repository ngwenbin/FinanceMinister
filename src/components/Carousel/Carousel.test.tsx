import { render, screen } from "@testing-library/react-native";
import "@testing-library/jest-native/extend-expect";

import Carousel from "./Carousel";
import { Text, View } from "react-native";

interface CarouselListData {
  id: string;
  title: string;
  subtitle: string;
}

const carouselListLength = 10;

const carouselList: Array<CarouselListData> = Array.from({
  length: carouselListLength,
}).map((_, i) => {
  return {
    id: `${i}`,
    title: `This is the title ${i + 1}!`,
    subtitle: `This is the subtitle ${i + 1}!`,
  };
});

const carouselContentLayout = (item: CarouselListData) => {
  return (
    <View
      testID={`carouselItem_${item.id}`}
      style={{
        width: 300,
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      <View style={{ width: "100%", height: 100 }} />
      <Text>{item.title}</Text>
      <Text>{item.subtitle}</Text>
    </View>
  );
};

describe("<Carousel>", () => {
  describe("as default", () => {
    test("Renders successfully", () => {
      const activeCarousel = render(
        <Carousel
          data={carouselList}
          contentRenderer={(item) => carouselContentLayout(item)}
          containerStyle={{ height: 400 }}
        />
      );
      expect(activeCarousel).toBeDefined();
    });

    test("Renders correctly", () => {
      const activeCarousel = render(
        <Carousel
          data={carouselList}
          contentRenderer={(item) => carouselContentLayout(item)}
          containerStyle={{ height: 400 }}
        />
      );
      expect(activeCarousel.toJSON()).toMatchSnapshot();
    });

    test("Outputs all child items", async () => {
      render(
        <Carousel
          data={carouselList}
          contentRenderer={(item) => carouselContentLayout(item)}
          containerStyle={{ height: 400 }}
          maxToRenderPerBatch={carouselListLength}
          initialNumToRender={carouselListLength}
        />
      );
      const items = await screen.findAllByTestId("carouselItem_", {
        exact: false,
      });
      expect(items).toHaveLength(carouselListLength);
    });
  });
});
