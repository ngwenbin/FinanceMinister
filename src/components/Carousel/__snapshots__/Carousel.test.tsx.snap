// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel> as default Renders correctly 1`] = `
<View
  style={
    Object {
      "height": 400,
    }
  }
>
  <RCTScrollView
    bounces={false}
    data={
      Array [
        Object {
          "id": "0",
          "subtitle": "This is the subtitle 1!",
          "title": "This is the title 1!",
        },
        Object {
          "id": "1",
          "subtitle": "This is the subtitle 2!",
          "title": "This is the title 2!",
        },
        Object {
          "id": "2",
          "subtitle": "This is the subtitle 3!",
          "title": "This is the title 3!",
        },
        Object {
          "id": "3",
          "subtitle": "This is the subtitle 4!",
          "title": "This is the title 4!",
        },
        Object {
          "id": "4",
          "subtitle": "This is the subtitle 5!",
          "title": "This is the title 5!",
        },
        Object {
          "id": "5",
          "subtitle": "This is the subtitle 6!",
          "title": "This is the title 6!",
        },
        Object {
          "id": "6",
          "subtitle": "This is the subtitle 7!",
          "title": "This is the title 7!",
        },
        Object {
          "id": "7",
          "subtitle": "This is the subtitle 8!",
          "title": "This is the title 8!",
        },
        Object {
          "id": "8",
          "subtitle": "This is the subtitle 9!",
          "title": "This is the title 9!",
        },
        Object {
          "id": "9",
          "subtitle": "This is the subtitle 10!",
          "title": "This is the title 10!",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    horizontal={true}
    initialNumToRender={0}
    keyExtractor={[Function]}
    maxToRenderPerBatch={1}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
    windowSize={2}
  >
    <View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "width": 300,
            }
          }
          testID="carouselItem_0"
        >
          <View
            style={
              Object {
                "height": 100,
                "width": "100%",
              }
            }
          />
          <Text>
            This is the title 1!
          </Text>
          <Text>
            This is the subtitle 1!
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "width": 6750,
          }
        }
      />
    </View>
  </RCTScrollView>
  <View
    pointerEvents="none"
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingVertical": 2.5,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#374151",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 7.5,
            "marginHorizontal": 2.5,
            "width": 7.5,
          },
          Object {
            "backgroundColor": "#D1D5DB",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;
